wicket-core/src/main/java/org/apache/wicket/resource/FileSystemResourceReference.java

```diff
@@ -0,0 +1,175 @@
+ * add(new Video("video", new FileSystemResourceReference(path)));
+ * </code>
+ * </pre>
+ * 
+ * @author Tobias Soloschenko
+ */
+public class FileSystemResourceReference extends ResourceReference
+{
+	private static final long serialVersionUID = 1L;
+
+	private Path path;
+
+	private static Map<URI, FileSystem> fileSystemURIs = new HashMap<URI, FileSystem>();
+
+	/**
+	 * Creates a file system resource reference based on the given path
+	 * 
+	 * @param name
+	 *            the name of the resource reference to expose data
+	 * @param path
+	 *            the path to create the resource reference
+	 */
+	public FileSystemResourceReference(String name, Path path)
+	{
+		super(name);
+		this.path = path;
+	}
+
+	/**
+	 * Creates a file system resource reference based on the given path
+	 * 
+	 * @param path
+	 *            the path to create the resource reference (the name is used to expose the data)
+	 */
+	public FileSystemResourceReference(Path path)
+	{
+		super(path.getFileName().toString());
+		this.path = path;
+	}
+
+	/**
+	 * Creates a new {@link org.apache.wicket.markup.html.media.FileSystemResource} and applies the
+	 * path to it.
+	 */
+	@Override
+	public IResource getResource()
+	{
+		return getFileSystemResource();
+	}
+
+	/**
+	 * Used to apply a custom mime type without implementing a mime type detection
+	 * 
+	 * @return the mime type
+	 * @throws IOException
+	 *             if the mime type could'nt be read
+	 */
+	protected String getMimeType() throws IOException
+	{
+		return getFileSystemResource().getMimeType();
+	}
+
+	/**
+	 * Gets the file system resource to be used for the resource reference
+	 * 
+	 * @return the file system resource to be used for the resource reference
+	 */
+	protected FileSystemResource getFileSystemResource()
+	{
+		return new FileSystemResource(path);
+	}
+
+	/**
+	 * Creates a path and a file system (if required) based on the given URI
+	 * 
+	 * @param uri
+	 *            the URI to create the file system and the path of
+	 * @param env
+	 *            the environment parameter to create the file system with
+	 * @return the path of the file in the file system
+	 * @throws IOException
+	 *             if the file system could'nt be created
+	 * @throws URISyntaxException
+	 *             if the URI has no valid syntax
+	 */
+	public static Path getPath(URI uri, Map<String, String> env) throws IOException,
+		URISyntaxException
+	{
+		String uriString = uri.toString();
+		int indexOfExclamationMark = uriString.indexOf('!');
+		if (indexOfExclamationMark == -1)
+		{
+			return Paths.get(uri);
+		}
+		String zipFile = uriString.substring(0, indexOfExclamationMark);
+		FileSystem fileSystem = fileSystemURIs.get(uri);
+		if (fileSystem == null)
+		{
+			if (env == null)
+			{
+				env = new HashMap<>();
+				env.put("create", "true");
+				env.put("encoding", "UTF-8");
+			}
+			fileSystem = FileSystems.newFileSystem(new URI(zipFile), env);
+			fileSystemURIs.put(uri, fileSystem);
```

Comment:
What is the reason of the caching of the file systems here?
It has some issues related to concurrency. Better use ConcurrentMap with `#putIfAbsent()`.
I'd prefer if Wicket doesn't cache it though. I can foresee some issues with redeployments and OSGi usage.